Bluetooth: host: Add initial support for Isochronous channels

Vudentz opened the pull request:
Very initial support for Isochronous Channels introduced in 5.2 which includes:

+ HCI command/event definitions
+ CIG/CIS/Unicast both as slave and master
+ Shell module
+ Update to drivers to support ISO packets (standard only)
- BIS/Broadcast is not supported (depends on periodic advertisements)

Vudentz committed:
Bluetooth: Add HCI PDU definitions for ISO channels

This adds the HCI commands and events defined as per Bluetooth 5.2
specification.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: Audio: Initial ISO channel support

This adds initial code for handling ISO channels.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: monitor: Add support for ISO packets

This adds types for TX/RX ISO packets so the likes of btmon can decode
them properly.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: H4: Add support for ISO packets

This adds supports for ISO packets so then can be transmitted and
received with H4 driver.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: userchan: Add support for ISO packets

This adds supports for ISO packets so then can be transmitted and
received with userchan driver.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: shell: Add CONFIG_BT_AUDIO

This enables use of iso module with shell application.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: hci_raw: Add support for ISO packets

This adds supports for ISO packets so then can be transmitted and
received with hci_raw driver.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: H5: Add support for ISO packets

This adds supports for ISO packets so then can be transmitted and
received with H5 driver.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

sjg20 commented on include/bluetooth/hci.h:
```diff
@@ -1258,6 +1322,213 @@ struct bt_hci_cp_le_set_privacy_mode {
 } __packed;
 
+#define BT_HCI_OP_LE_READ_BUFFER_SIZE_V2        BT_OP(BT_OGF_LE, 0x60)
+struct bt_hci_rp_le_read_buffer_size_v2 {
```
Why no comments?

Vudentz replied:
What comments are you expecting? There is no comments to any HCI commands in this files since they are a direct map to the specification.

sjg20 commented on include/bluetooth/iso.h:
```diff
@@ -0,0 +1,256 @@
+	struct bt_iso_chan_ops		*ops;
+	/** Channel QoS reference */
+	struct bt_iso_chan_qos		*qos;
+	/** Channel data path reference*/
```
comment style

Vudentz replied:
Can you be more specific? CI didn't catch any documentation style problem in these.

sjg20 commented on include/bluetooth/iso.h:
```diff
@@ -0,0 +1,256 @@
+	uint8_t				sca;
+	/** Channel packing mode */
+	uint8_t				packing;
+	/** Channel framing mode */
```
But what does it mean? E.g. should I set it to 23?

In general these comments don't seem to stand on their own

Vudentz replied:
It is the framing mode from the spec, either framed or unframed, the upper layer will have to decide in which mode it shall be used.

sjg20 commented on subsys/bluetooth/host/audio/Kconfig:
```diff
@@ -0,0 +1,109 @@
+	default 1
+	range 1 255
+	help
+	  Number of buffers available for incoming Isochronous channel packets.
```
Guidance on when to change this?

Vudentz replied:
This should usually be set with the board definition, much like BT_ACL_RX_COUNT.

sjg20 commented on subsys/bluetooth/host/audio/iso_internal.h:
```diff
@@ -0,0 +1,111 @@
+ */
+
+/*
+ * Copyright (c) 2018 Intel Corporation
```
Is this old code?

Vudentz replied:
Will fix it.

sjg20 commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -5432,16 +5445,71 @@ static void le_read_buffer_size_complete(struct net_buf *buf)
+
+	bt_dev.le.iso_mtu = sys_le16_to_cpu(rp->iso_mtu);
+	if (!bt_dev.le.iso_mtu) {
+		return;
```
Is this an error? Can you return it, or signal it somehow?

Emil-Gydesen-Bose replied:
Or perhaps at least have `BT_DBG` or higher level log statement

sjg20 commented on subsys/bluetooth/shell/iso.c:
```diff
@@ -0,0 +1,256 @@
+
+	if (iso_chan.conn) {
+		printk("No channels available\n");
+		return -ENOMEM;
```
-ENOSPC?

sjg20 commented on subsys/bluetooth/shell/iso.c:
```diff
@@ -0,0 +1,256 @@
+	}
+
+	if (iso_chan.conn) {
+		shell_error(shell, "Already binded");
```
bound

Vudentz replied:
Will fix it.

sjg20 commented on subsys/bluetooth/shell/iso.c:
```diff
@@ -0,0 +1,256 @@
+	}
+
+	if (!iso_chan.conn) {
+		shell_error(shell, "Not binded");
```
bound

I don't think binded is a word

sjg20 commented on subsys/bluetooth/shell/iso.c:
```diff
@@ -0,0 +1,256 @@
+		if (ret < 0) {
+			shell_print(shell, "Unable to send: %d", -ret);
+			net_buf_unref(buf);
+			return -ENOEXEC;
```
Strange error. Why not return ret?

Also since this is a command, don't you need to return 0?

I don't know much about Zephyr though

Vudentz replied:
See the shell for this, not to be rude or anything but it is more productive to only comment on things you actually have knowlodge that it might be a problem, or perhaps look for other shell modules such as bt.c where it basically does the same thing regarding the use of -ENOEXEC to indicate the command could not be performed.

sjg20 commented on subsys/bluetooth/host/monitor.h:
```diff
@@ -24,6 +24,8 @@
 #define BT_MONITOR_VENDOR_DIAG	11
 #define BT_MONITOR_SYSTEM_NOTE	12
 #define BT_MONITOR_USER_LOGGING	13
+#define BT_MONITOR_ISO_TX_PKT	18
```
Line up with those above?

Vudentz replied:
This file is using tab to align these, I will convert it to spaces since github don't seem to be able to format tabs properly.

Emil-Gydesen-Bose commented on subsys/bluetooth/host/audio/Kconfig:
```diff
@@ -0,0 +1,109 @@
+
+config BT_AUDIO_DEBUG_ISO
+	bool "ISO channel debug"
+	depends on BT_DEBUG
```
Superfluous  `depends on` as this requires `BT_AUDIO_DEBUG` to be enabled which depends on `BT_DEBUG`.

Vudentz replied:
Sure, will fix it.

Emil-Gydesen-Bose commented on subsys/bluetooth/host/audio/iso.c:
```diff
@@ -0,0 +1,1075 @@
+
+void hci_le_cis_estabilished(struct net_buf *buf)
+{
+	struct bt_hci_evt_le_cis_established *evt = (void *)buf->data;
```
Shouldn't the cast of the buf->data be of type `struct bt_hci_evt_le_cis_established *` instead of `void *`?

Vudentz replied:
I would guess void * is fine here since the equivalent net_buf_pull also would result in void * being returned.

Emil-Gydesen-Bose commented on subsys/bluetooth/host/audio/iso.c:
```diff
@@ -0,0 +1,1075 @@
+			return -ENOMEM;
+		}
+
+		/* TODO: Figure out if ID was already assigned? */
```
Can't this be done with a simple for loop over existing ISOs?

Vudentz replied:
What this is doing is allocating a ISO (bt_conn) object for each ACL since that is then used by hci_le_set_cig_params, so there is no existing ISOs yet, or the user want to create new connections anyway.

Emil-Gydesen-Bose replied:
If there are no existing ISOs yet, do we need to check if IDs was assigned here? In other words, can the TODO be removed?

Vudentz replied:
The TODO is for the CIG id which may have been set by the application so I think we will need to lookup if there is already in use. I do intend to add auto allocate logic though, in case the application don't care to use a specific CIG id.

Emil-Gydesen-Bose commented on subsys/bluetooth/host/audio/Kconfig:
```diff
@@ -0,0 +1,109 @@
+if BT_AUDIO_UNICAST
+
+config BT_MAX_ISO_CONN
+	int "Maximum number of simultaneous ISO connections"
```
Is the in any way depending on `CONFIG_BT_MAX_CONN`?

Vudentz replied:
I guess the minimal would have to be based on CONFIG_BT_MAX_CONN, but we can end up having more than one ISO per ACL pair.

Emil-Gydesen-Bose replied:
I'd suggest to set the minimum to CONFIG_BT_MAX_CONN, so that we can catch mis-configurations at compile times rather than runtime. If there's a maxmium number of ISO per ACL, then I'd suggest also using CONFIG_BT_MAX_CONN times the max number of ISO per ACL as the maximum here

Vudentz replied:
Done.

Emil-Gydesen-Bose commented on subsys/bluetooth/host/audio/iso.c:
```diff
@@ -0,0 +1,1075 @@
+
+	memset(&param, 0, sizeof(param));
+
+	/* TODO: Allow setting ID? */
+	param.id = 0x01;
```
Not sure if the API should allow setting the ID, but shouldn't the ID at least be incremented for each call, or something?

Vudentz replied:
It should, or actually have it allocated with connection index or something like it, I will fix it.

Emil-Gydesen-Bose commented on subsys/bluetooth/host/audio/iso.c:
```diff
@@ -0,0 +1,1075 @@
+
+void bt_iso_recv(struct bt_conn *conn, struct net_buf *buf, uint8_t flags)
+{
+	struct bt_hci_iso_data_hdr *hdr;
```
Do we need to check or handle the `sn` in this function? Seems to be largely ignored.

Vudentz replied:
Good point, I will need to check if the sn is there to actually tell for which packet it is, though I don't think we can get multiple outstanding packets being reassembled so I guess we could just compare if the sn matches or not, and if doesn't discard the all fragments.

Emil-Gydesen-Bose replied:
Agreed

Emil-Gydesen-Bose commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -40,12 +40,14 @@
 #include "crypto.h"
 #include "settings.h"
 
+
```
Superfluous newline

Vudentz replied:
Will fix it.

Emil-Gydesen-Bose reviewed (commented):
A few comments/questions, otherwise it generally LGTM.

Thalley commented on subsys/bluetooth/host/audio/iso.c:
```diff
@@ -0,0 +1,1075 @@
+	return bt_hci_cmd_send_sync(BT_HCI_OP_LE_REMOVE_CIG, buf, NULL);
+}
+
+void bt_iso_cleanup(struct bt_conn *conn)
```
This function seems to be valid for CIS only. Perhaps rename to `bt_iso_cleanup_cis` or use handle as argument instead of a `bt_conn`?

Vudentz replied:
Will change it to just take the handle, I guess any command which applies to BIS as well will need to map to handle, we could in theory make a BIS map to a bt_conn as well, since the ISO packets are also meant to carry broadcast data.

Thalley replied:
> we could in theory make a BIS map to a bt_conn as well
I was wondering the same thing. Would be great if we somehow could use the same functions for sending data over BIS as we do for CIS.

Thalley commented on subsys/bluetooth/host/audio/iso.c:
```diff
@@ -0,0 +1,1075 @@
+	return 0;
+}
+
+static int hci_le_setup_iso_data_path(struct bt_conn *conn,
+				      struct bt_iso_data_path *path)
```
It seems that the `bt_conn` argument is only used for the handle. Since BIS also use this command, I think the argument should be changed to a `uint16_t handle` rather than a `bt_conn` so that it applies for both CIS and BIS. 

Seems to also be existing for more functions in this file.

Vudentz replied:
I leaning towards making BIS map to a bt_conn as well since it is effectively an ISO connection we probably want to reuse the same code.

Vudentz committed:
Bluetooth: shell: Add iso module

This adds iso shell module which we can be used to exercise ISO related
APIs:

iso - Bluetooth ISO shell commands
Subcommands:
  bind        :[dir] [interval] [packing] [framing] [latency] [sdu]
  [phy] [rtn]
  connect     :Connect ISO Channel
  listen      :[security level]
  send        :Send to ISO Channel
  disconnect  :Disconnect ISO Channel

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Thalley commented on subsys/bluetooth/host/audio/Kconfig:
```diff
@@ -0,0 +1,108 @@
+
+if BT_AUDIO
+
+config BT_ISO
+	# Virtual/hidden option
+	bool
+
+if BT_CONN
+
+config BT_AUDIO_UNICAST
+	bool "Bluetooth Unicast Audio Support"
+	select BT_SMP
+	select BT_L2CAP_DYNAMIC_CHANNEL
+	select BT_ISO
+	select BT_GATT_DYNAMIC_DB
+	select BT_GATT_CACHING
+	select BT_EATT
+	help
+	  This option enables support for Bluetooth Unicast Audio using
+	  Isochronous channels.
```
I realize that the only (spec) defined use case of ISO is for audio, but I think we should consider decoupling audio and ISO, as ISO may have other use cases than just audio, and it would be somewhat weird having to enable audio in order to use ISO for non-audio use cases.

Vudentz replied:
Yep, but for now I decided to leave it under audio to make it simple, if in the future there are other users we can decouple the options and perhaps have a dedicated directiory for iso or just its handling to the directory bellow.

Vudentz commented on include/bluetooth/hci.h:
```diff
@@ -59,6 +67,45 @@ struct bt_hci_acl_hdr {
+
+#define bt_iso_handle(h)                ((h) & 0x0fff)
+#define bt_iso_flags(h)                 ((h) >> 12)
+#define bt_iso_flags_pb(f)              (f & 0x0003)
+#define bt_iso_flags_ts(f)              ((f >> 2) & 0x0001)
+#define bt_iso_pack_flags(pb, ts)       ((pb & 0x0003) | ((ts & 0x0001) << 2))
```
Done.

Vudentz commented on subsys/bluetooth/host/audio/Kconfig:
```diff
@@ -0,0 +1,108 @@
+	  and there are no dedicated fragment buffers, a deadlock may occur.
+	  In most cases the default value of 2 is a safe bet.
+
+config BT_ISO_TX_MTU
+	int "Maximum supported MTU for Isochronous TX buffers"
+	default 251
+	range 23 2000
+	help
+	  Maximum MTU for Isochronous channels TX buffers.
+
+config BT_ISO_RX_BUF_COUNT
+	int "Numer of Isochronous RX buffers"
+	default 1
+	range 1 255
+	help
+	  Number of buffers available for incoming Isochronous channel packets.
+
+config BT_ISO_RX_MTU
+	int "Maximum supported MTU for Isochronous RX buffers"
+	default 251
+	range 23 2000
+	help
+	  Maximum MTU for Isochronous channels RX buffers.
```
These are the buffers for ISO connections, so yes it should apply for both CIS and BIS.

Vudentz commented on subsys/bluetooth/host/audio/iso.c:
```diff
@@ -0,0 +1,1074 @@
+	for (i = 0; i < param->num_conns; i++) {
+		conn = param->conns[i];
+
+		if (conn->type != BT_CONN_TYPE_LE) {
+			return -EINVAL;
+		}
+
+		conn = bt_conn_add_iso(conn);
+		if (!conn) {
+			return -ENOMEM;
+		}
```
Done.

Vudentz commented on subsys/bluetooth/host/audio/Kconfig:
```diff
@@ -0,0 +1,108 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+menuconfig BT_AUDIO
+	bool "Bluetooth Audio support"
+	help
+	  This option enables Bluetooth Audio support. The specific
+	  features that are available may depend on other features
+	  that have been enabled in the stack, such as Periodic
+	  Advertisement for Broadcast and L2CAP Dynamic Channel
+	  for Unicast.
```
Yep, update it with [EXPERIMENTAL]

Vudentz commented on subsys/bluetooth/host/conn_internal.h:
```diff
@@ -190,6 +204,26 @@ bool bt_conn_exists_le(uint8_t id, const bt_addr_le_t *peer);
+struct bt_conn *bt_conn_add_iso(struct bt_conn *acl);
+
+/* Cleanup ISO references */
+void bt_iso_cleanup(struct bt_conn *sco_conn);
```
Typo, it should be iso_conn there.

Vudentz commented on subsys/bluetooth/host/audio/iso_internal.h:
```diff
@@ -0,0 +1,111 @@
+void hci_le_cis_req(struct net_buf *buf);
+
+/* Notify ISO channels of a new connection */
+int bt_iso_accept(struct bt_conn *conn);
+
+/* Notify ISO channels of a new connection */
+void bt_iso_connected(struct bt_conn *conn);
+
+/* Notify ISO channels of a disconnect event */
+void bt_iso_disconnected(struct bt_conn *conn);
```
Those in iso.h operate on ISO connections only, including bt_iso_accept, that said I can make it check for proper types, these are internal APIs so I don't expect these to be used directly to cause this sort of confusion.

Thalley reviewed (approved):
LGTM

Thalley commented on subsys/bluetooth/host/conn_internal.h:
```diff
@@ -160,6 +169,9 @@ struct bt_conn {
 		struct bt_conn_sco	sco;
+#endif
+#if defined(CONFIG_BT_AUDIO)
+		struct bt_conn_iso	iso;
```
I'm not allowed to add a comment on a unchanged line: It seems that `bt_conn` `channels` field is also used for `bt_iso_chan` and not just L2CAP channels, so perhaps the comment on line 160 should be updated?

Vudentz replied:
Yep, just fixed that.

carlescufi commented:
This needs to go in before the 2.4 feature merge window closes. @joerchan and @jhedberg could you review, even if it's not in-depth?

jhedberg commented on drivers/bluetooth/hci/h4.c:
```diff
@@ -87,6 +89,13 @@ static inline void h4_get_type(void)
+			rx.hdr_len = rx.remaining;
+			break;
+		}
+	/* Fallthrough */
```
Is this still alright? I remember seeing some changes to this annotation in the tree.

Vudentz replied:
I guess you are talking about the use of __fallthrough, I just converted these comments to use that instead.

jhedberg commented on drivers/bluetooth/hci/h4.c:
```diff
@@ -307,6 +337,12 @@ static inline void read_header(void)
+			get_iso_hdr();
+			break;
+		}
+	/* Fallthrough */
```
Same here

jhedberg commented on drivers/bluetooth/hci/h4.c:
```diff
@@ -344,6 +380,12 @@ static inline void process_tx(void)
+				tx.type = H4_ISO;
+				break;
+			}
+		/* Fallthrough */
```
And here

jhedberg commented on drivers/bluetooth/hci/userchan.c:
```diff
@@ -57,11 +58,21 @@ static int bt_dev_index = -1;
+		if (IS_ENABLED(CONFIG_BT_ISO)) {
+			return bt_buf_get_rx(BT_BUF_ISO_IN, K_FOREVER);
+		}
+		/* Fallthrough */
```
Same here

jhedberg commented on drivers/bluetooth/hci/userchan.c:
```diff
@@ -131,6 +142,12 @@ static int uc_send(struct net_buf *buf)
+			net_buf_push_u8(buf, H4_ISO);
+			break;
+		}
+		/* Fallthrough */
```
And here

jhedberg commented on subsys/bluetooth/host/hci_raw.c:
```diff
@@ -171,6 +186,12 @@ int bt_recv(struct net_buf *buf)
+				net_buf_push_u8(buf, H4_ISO);
+				break;
+			}
+		/* Fallthrough */
```
And here

justinleavesley commented:
I really hope this can be merged for 2.4 @jhedberg

jhedberg commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1353,6 +1456,57 @@ static void process_unack_tx(struct bt_conn *conn)
+			continue;
+		}
+
+		if (&iso_conns[i] == conn) {
+			return conn;
+		}
+
+		if (iso_conns[i].iso.acl == conn) {
+			return bt_conn_ref(&iso_conns[i]);
+		}
```
This looks rather suspicious and seems like bad design: in some cases the function returns a new reference and in other cases it doesn't.

Vudentz replied:
Nice catch, I seems it doesn indeed expect a new ref in all cases, I will fix it.

joerchan commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -5245,16 +5257,72 @@ static void le_read_buffer_size_complete(struct net_buf *buf)
+
+	return bt_hci_cmd_send_sync(BT_HCI_OP_LE_SET_HOST_FEATURE, buf, NULL);
+#else
+	return 0;
```
-ENOTSUP?

Vudentz replied:
Will fix it.

joerchan commented on subsys/bluetooth/host/hci_core.h:
```diff
@@ -180,8 +180,12 @@ struct bt_dev_le {
 	struct k_sem		pkts;
+	uint16_t		acl_mtu;
+	struct k_sem		acl_pkts;
+	uint16_t		iso_mtu;
```
ifdef iso_mtu and iso_pkts?

Vudentz replied:
I guess you are suggesting to make iso related parameters conditional to CONFIG_BT_ISO? Or is it something else?

joerchan replied:
I'm suggesting to wrap iso_mtu and iso_pkts under `#if defined(CONFIG_BT_ISO)` in this struct.

joerchan commented on subsys/bluetooth/host/conn_internal.h:
```diff
@@ -207,14 +241,29 @@ struct bt_conn *bt_conn_lookup_addr_br(const bt_addr_t *peer);
+		    conn->type == BT_CONN_TYPE_ISO) {
+			return true;
+		}
+	/* Fall-through */
```
`__fallthrough`?

Vudentz replied:
Will fix.

joerchan commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1353,6 +1456,57 @@ static void process_unack_tx(struct bt_conn *conn)
+	return NULL;
+}
+
+#if defined(CONFIG_BT_ISO)
```
Move the `#if` into the function, can avoid  the #else by having the `return NULL;` at the end.

Vudentz replied:
Fixed.

joerchan commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1214,41 +1296,62 @@ static void conn_cleanup(struct bt_conn *conn)
 }
 
+static int conn_prepare_events(struct bt_conn *conn,
+			       struct k_poll_event events[])
```
Only given one entry, use `struct k_poll_event *events`?

Vudentz replied:
Fixed.

Vudentz commented on include/bluetooth/hci.h:
```diff
@@ -1277,6 +1343,213 @@ struct bt_hci_cp_le_set_privacy_mode {
 	uint8_t         mode;
 } __packed;
 
+#define BT_HCI_OP_LE_READ_BUFFER_SIZE_V2        BT_OP(BT_OGF_LE, 0x60)
```
Will fix them.

Vudentz commented on include/bluetooth/hci.h:
```diff
@@ -59,6 +67,46 @@ struct bt_hci_acl_hdr {
 } __packed;
 #define BT_HCI_ACL_HDR_SIZE             4
 
+#define BT_ISO_START                    0x00
```
These are following BT_ACL headers, but Im fine changing them.

joerchan replied:
I noticed that as well. The rest is prefixed bt_hci_iso, so bt_acl is kind of the odd one out.
Whichever you decide is fine with me.

Vudentz commented on include/bluetooth/hci.h:
```diff
@@ -22,6 +22,14 @@
+/* Company Identifiers (see Bluetooth Assigned Numbers) */
+#define BT_COMP_ID_LF           0x05f1
+
+#define BT_ADDR_LE_PUBLIC       0x00
```
Weird, that must have been included during the rebase somehow.

Vudentz commented on include/bluetooth/hci.h:
```diff
@@ -22,6 +22,14 @@
 #endif
 
+/* Company Identifiers (see Bluetooth Assigned Numbers) */
+#define BT_COMP_ID_LF           0x05f1
```
Will fix it.

Vudentz commented on drivers/bluetooth/hci/h4.c:
```diff
@@ -307,6 +337,12 @@ static inline void read_header(void)
+			get_iso_hdr();
+			break;
+		}
+		__fallthrough;
```
Weird, I thought that should be printing an error no CODE_UNREACHABLE, even the original code is wrong is that regard since for other code path we just print an error and continue.

Vudentz replied:
Actually the code is correct, we can't continue if we cannot process the header, so either we continue using CODE_UNREACHABLE or we do process the header and drop it later, but given that ISO never to be enabled by the stack there is really something wrong if we got an ISO packet. @jhedberg since you are the original author do you have any opnion on this? Shall we continue using CODE_UNREACHABLE if we cannot parse the header (disabled CONFIG_BT_ISO) or just parse it and drop it later?

joerchan replied:
Johan won't be able to respond before the feature freeze. It just looks weird but I understand the logic. 
I suppose the CODE_UNREACHABLE is not needed, since there is a return after there. Keeping it or removing it, either is fine with me.

Vudentz replied:
I think the idea was to cause __builtin_unreachable which seems to be used by fatal errors, perhaps it would be a better idea to assert in that case.

joerchan commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1214,41 +1296,62 @@ static void conn_cleanup(struct bt_conn *conn)
+
+	BT_DBG("Adding conn %p to poll list", conn);
+
+	k_poll_event_init(&events[0],
```
could be just `events` here.

and `event->tag` further below.

Vudentz replied:
Right, that is something that we can indeed improve latter on.

joerchan reviewed (approved):
LGTM, left a few comments that could be considered.
